### PR TITLE
feat(transaction-pool): make `EthTransactionValidator` generic over Validator

### PR DESCRIPTION
**Pull request by tcoratger**
Should solve #4255.

Not sure moving `TransactionValidationTaskExecutor` and `EthTransactionValidator` structs to `task.rs` is a good idea because this will induce multiple duplicate imports and some changes (for example `EthTransactionValidatorBuilder` must be exposed publicly in order to be accessible in `task.rs`), what do you think?

### CODE REVIEW

File: crates/transaction-pool/src/validate/eth.rs [hunk 1]
```diff
@@ -22,27 +22,34 @@ use tokio::sync::{oneshot, Mutex};
...
-    to_validation_task: Arc<Mutex<ValidationJobSender>>,
 }
 
-// === impl EthTransactionValidator ===
+// === impl TransactionValidationTaskExecutor ===
 
-impl EthTransactionValidator<(), ()> {
+impl TransactionValidationTaskExecutor<()> {
     /// Convenience method to create a [EthTransactionValidatorBuilder]
     pub fn builder(chain_spec: Arc<ChainSpec>) -> EthTransactionValidatorBuilder {
```
Comment (mattsse) [inline crates/transaction-pool/src/validate/eth.rs:43]:
```suggestion
    pub fn eth_builder(chain_spec: Arc<ChainSpec>) -> EthTransactionValidatorBuilder {
```

File: crates/transaction-pool/src/validate/eth.rs [hunk 2]
```diff
@@ -22,27 +22,34 @@ use tokio::sync::{oneshot, Mutex};
...
         EthTransactionValidatorBuilder::new(chain_spec)
     }
 }
 
-impl<Client, Tx> EthTransactionValidator<Client, Tx> {
+impl<Client, Tx> TransactionValidationTaskExecutor<EthTransactionValidator<Client, Tx>> {
```
Comment (mattsse) [inline crates/transaction-pool/src/validate/eth.rs:48]: this should be generic over V: TransactionValidator + Clone instead

**Review by mattsse — CHANGES_REQUESTED** [review]
thanks, this is almost how I envisioned this.

I'd like to impl Validator for Task over V so it isn't tied to the eth validator anymore

File: crates/transaction-pool/src/validate/eth.rs [hunk 3]
```diff
@@ -22,27 +22,34 @@ use tokio::sync::{oneshot, Mutex};
...
     /// Creates a new instance for the given [ChainSpec]
     ///
     /// This will spawn a single validation tasks that performs the actual validation.
-    /// See [EthTransactionValidator::with_additional_tasks]
+    /// See [TransactionValidationTaskExecutor::with_additional_tasks]
     pub fn new<T>(client: Client, chain_spec: Arc<ChainSpec>, tasks: T) -> Self
```
[the remark this answers is not in the document]
Comment (tcoratger) [reply crates/transaction-pool/src/validate/eth.rs:53]: @mattsse Not sure I understand due to the previous comment, if the implementation is generic over `V: TransactionValidator + Clone`, how can we obtain the `Client` type to construct the `new` function (which is now renamed to `eth`)?

Also, the `fn new(V)` is additional function, meaning a 3rd function to add here in your mind?

[the remark this answers is not in the document]
Comment (mattsse) [reply crates/transaction-pool/src/validate/eth.rs:53]: ah yeah, sorry I wasn't really clear on this, the client and chainspec fields are only needed for constructing the eth validator.

we can keep these function as a convenience for ` impl TransactionValidationTaskExecutor<EthTransactionValidator<...>> {fn eth(...)}`

but everything else should be generic over V

> Also, the fn new(V) is additional function, meaning a 3rd function to add here in your mind?

yes in `impl<V> TransactionValidationTaskExecutor<V> {fn new(validator: V) -> Self}`

File: crates/transaction-pool/src/validate/eth.rs [hunk 4]
```diff
@@ -22,27 +22,34 @@ use tokio::sync::{oneshot, Mutex};
 ///
 /// This validator is non-blocking, all validation work is done in a separate task.
 #[derive(Debug, Clone)]
+pub struct TransactionValidationTaskExecutor<V> {
+    /// The validator that will validate transactions on a separate task.
+    validator: V,
+    /// The sender half to validation tasks that perform the actual validation.
+    to_validation_task: Arc<Mutex<ValidationJobSender>>,
+}
```
[the remark this answers is not in the document]
Comment (tcoratger) [reply crates/transaction-pool/src/validate/eth.rs]: @mattsse Done

File: crates/transaction-pool/src/validate/eth.rs (hunk 2 above)
Comment (tcoratger) [reply crates/transaction-pool/src/validate/eth.rs]: @mattsse Done

**Review by mattsse — APPROVED** [review]
amazing ty!

**Pull request merged**